primitives/transaction/src/signature_proof.rs

```diff
@@ -0,0 +1,512 @@
+
+        // Setup inner fields
+        let mut client_data_flags = WebauthnClientDataFlags::default();
+        let mut client_data_extra_fields = "".to_string();
+
+        // Convert client_data_json bytes to string for search & split operations
+        // FIXME: Handle invalid UTF-8
+        let client_data_json_str = std::str::from_utf8(client_data_json).unwrap();
+
+        // Extract origin from client_data_json
+        let origin_search_term = r#","origin":""#;
+        // Find the start of the origin value
+        // FIXME: Handle missing origin
+        let origin_start =
+            client_data_json_str.find(origin_search_term).unwrap() + origin_search_term.len();
+        // Find the closing quotation mark of the origin value
+        // FIXME: Handle missing closing quotation mark
+        let origin_length = client_data_json_str[origin_start..].find('"').unwrap();
+        // The origin is the string between the two indices
+        let origin = &client_data_json_str[origin_start..origin_start + origin_length];
+
+        // Compute and compare RP ID with authenticatorData
+        let url = url::Url::parse(origin)?;
+        let hostname = url.host_str().unwrap(); // FIXME: Handle missing hostname
+        let rp_id = nimiq_hash::Sha256Hasher::default().digest(hostname.as_bytes());
+        if rp_id.0 != authenticator_data[0..32] {
+            return Err(SerializationError::new(
+                "Computed RP ID does not match authenticator data",
+            ));
+        }
+
+        // Compute host field, which includes the port if non-standard
+        let port_suffix = if let Some(port) = url.port() {
+            format!(":{}", port)
+        } else {
+            "".to_string()
+        };
+        let host = format!("{}{}", hostname, port_suffix);
+
+        // Check if client_data_json contains any extra fields
+        // Search for the crossOrigin field first
+        let parts = client_data_json_str
+            .split(r#""crossOrigin":false"#)
+            .collect::<Vec<_>>();
+        let suffix = if parts.len() == 2 {
+            parts[1]
+        } else {
+            // Client data does not include the crossOrigin field
+            client_data_flags.insert(WebauthnClientDataFlags::NO_CROSSORIGIN_FIELD);
+
+            // We need to check for extra fields after the origin field instead
+            let parts = client_data_json_str
+                .split(&format!(r#""origin":"{}""#, origin))
+                .collect::<Vec<_>>();
+            assert_eq!(parts.len(), 2);
+            parts[1]
+        };
+
+        // Check if the suffix contains extra fields
+        if suffix.len() > 1 {
+            // Cut off first comma and last curly brace
+            client_data_extra_fields = suffix[1..suffix.len() - 1].to_string();
+        }
+
+        if origin.contains(r":\/\/") {
+            client_data_flags.insert(WebauthnClientDataFlags::ESCAPED_ORIGIN_SLASHES);
+        }
```

Comment:
Consider using a JSON parser here instead of parsing `client_data_json` by hand. Maybe put the function behind a feature flag to not carry the dependency along everywhere. At least we should take care of the `FIXME`s to not crash the client if the caller sends junk.

Reply:
The problem with JSON parsers is that they delete the backward-slash that escapes the forward slash in Android Chrome's `origin` field (`https:\/\/`), which are nonetheless relevant to the signature! So when they are removed, it becomes impossible to know if they should be there or not, making signatures from Android Chrome invalid.

I can of course keep searching for a parser that handles it "correctly", or fork one, but since the JSON in client_data_JSON is otherwise standardized, I think we can also parse it manually.